src/proto/h1/conn.rs

```diff
@@ -545,6 +552,9 @@ where I: AsyncRead + AsyncWrite,
     // - Client: there is nothing we can do
     // - Server: if Response hasn't been written yet, we can send a 4xx response
     fn on_parse_error(&mut self, err: ::Error) -> ::Result<()> {
+        if self.shutdown_on_error == false {
```

Comment:
This could probably be inside `T::on_error` instead. Hm, it could perhaps be useful if the `on_error` method were adjusted to allow accessing the current `&read_buf`. Then, `ServerTransaction::on_error` could check to see if it at least starts with the HTTP/2 magic prefix before deciding to bubble the error back up (and let `server::Connection` try h2), or to just return the HTTP/1 bad response.